### DISCUSSION TITLE
Pixi and ROS2 Jazzy

### DISCUSSION THREAD

**Discussion by taylorpool** [prefix-dev/pixi#1524] [Q&A] [answered]
Hello,

This looks like an amazing project. I saw that there was a channel dedicated to ROS2 Humble. ROS2 Jazzy is the newest LTS version and represents a large step forward. What would it take for me to create a channel representing ROS2 Jazzy, or to add it to an existing channel?

Once again, thanks for all your great work!

**Comment by taylorpool** [comment, marked as answer]
Found the answer to my own question in the Robostack Paper: https://arxiv.org/pdf/2104.12910

**Reply by wolfv** [reply]
@taylorpool I'll start a `ros-jazzy` repository tomorrow - and then we can start building some new packages :)

**Reply by wolfv** [reply]
It will look more-or-less like this: https://github.com/RoboStack/ros-humble

**Reply by taylorpool** [reply]
Sounds great! Lmk how I can help!

**Reply by wolfv** [reply]
![image](https://github.com/prefix-dev/pixi/assets/885054/f13293a1-bfa7-4cab-aa0a-679931365099)

I got this to work! :) I'll try to enable the CI pipelines one of these days and then we can work our way to build all packages.

**Reply by wolfv** [reply]
The repository is here: https://github.com/RoboStack/ros-jazzy

**Reply by AndBondStyle** [reply]
@wolfv our team is also looking for an alternative to rosdep (which doesn't support version pinning and caused us a lot of pain recently). However we're based on ROS2 Iron and there aren't any channel for it yet. Although it would be nice for pixi registry to have channels for multiple versions of ROS (or at least those that [hasn't reached EOL yet](https://docs.ros.org/en/rolling/Releases.html)), I understand that it may be a hassle to maintain all of them (unless you automate it somehow?).

So the way to go for us is to create a private channel then? Or can you spin up another public repo for Iron too? And can you point me to where I can learn about repackaging ROS2 stuff?

**Reply by wolfv** [reply]
Hi @AndBondStyle - thanks for dropping by! If you are interested, I would be happy to set up a call (30 min or so) to go over our infrastructure. A lot of it is in fact automated (using a tool called `vinca`). 

It's indeed more work to support multiple releases and we are a relatively small team (mostly @Tobias-Fischer, @traversaro, me and a few others). For `jazzy` I would like to revamp the infrastructure to use `rattler-build` and some other tweaks that will hopefully make things easier going forward :)

**Reply by traversaro** [reply]
FYI if you are really interested in Iron, there was an attempt on starting a iron channel from @IsabelParedes, you can find some details in https://github.com/RoboStack/robostack.github.io/issues/24 and https://github.com/IsabelParedes/ros-iron .

**Reply by AndBondStyle** [reply]
@wolfv I think it would be more convenient to just record a video instead of a call, or make some sort of a draft wiki page going over the infrastructure, so we can share it later with others. Thank you for keywords, I'll start digging into vinca and rattler-build.

@traversaro sorry for a dumb question, but how do github repos (including [IsabelParedes/ros-iron](https://github.com/IsabelParedes/ros-iron)) correlate to what I see on https://prefix.dev/channels?

**Reply by traversaro** [reply]
> @traversaro sorry for a dumb question, but how do github repos (including [IsabelParedes/ros-iron](https://github.com/IsabelParedes/ros-iron)) correlate to what I see on https://prefix.dev/channels?

Not dumb at all, and given the lack of complete docs, asking questions is the best way to understand what's going on. Each GitHub repository stores the configuration files for vinca, the tool used to generate conda-build (and soon rattler-build) recipes from the rosdistro metadata. For each new commit in the repo, the GitHub Action https://github.com/RoboStack/ros-humble/blob/main/.github/workflows/main.yml checks for change in files, and generate new recipes if necessary on the `buildbranch_<platform>`, see for example the commit https://github.com/RoboStack/ros-humble/commit/cd66bb058abfbd3bb56d35dd39868f6776aa689f in https://github.com/RoboStack/ros-humble/tree/buildbranch_linux . As part of the commit, a GitHub Action job is generated (for example in that commit is the file `.github/workflows/linux.yml`) that then runs and actually compiles the packages that then are uploaded over at https://anaconda.org/RoboStack-staging, see https://github.com/RoboStack/ros-humble/actions/runs/9616368683/job/26525724034 . 

The channel https://anaconda.org/RoboStack-staging is then mirrored over at https://prefix.dev/channels/robostack-staging, but I am not familiar with that process, probably @wolfv knows more?

**Reply by AndBondStyle** [reply]
@traversaro that explains a lot, thank you. So if I want to start experimenting with ROS Iron packages, for example, I need to:
1. Fork a repo from [IsabelParedes/ros-iron](https://github.com/IsabelParedes/ros-iron) (or maybe create an fresh one with only necessary stuff)
2. Create a conda channel and API token for publishing
3. Make some changes and let CI do its thing

And then how do I get pixi to see and install those conda packages?

**Reply by wolfv** [reply]
Hi @AndBondStyle - sure we can also do that! :) It would probably be nice to have more complete documentation for the whole RoboStack stuff. 

If you want to follow the more traditional way, you would create a "channel" on anaconda.org, and generate an API key there that you can use for uploading. The RoboStack repositories also need a second Github Token to be able to _push_ the generated Github Actions workflows to a branch on the same repository (`buildbranch_linux`, `buildbranch_win`, etc.)

If you want to install the packages with `pixi`, you can just add:

```
channels = ["mychannel", "conda-forge"]
```

where `mychannel` would be the name of your channel. If the channel was on `prefix.dev`, then you would use `["https://prefix.dev/mychannel", ...]` etc. - currently `anaconda.org` is the "default namespace".

**Reply by traversaro** [reply]
Yes, that is basically it, even if probably for debugging you may want to build locally for faster feedback loops. `pixi` by default can install packages from all conda channels from anaconda.org, even if they are not mirrored on prefix.dev (to do a check, you can install `icub-basic-demos` from the `robotology` channel, that is on anaconda but is is not mirrored in prefix.dev).

**Reply by wolfv** [reply]
Yes, when you build packages locally, you create a "local" conda channel (a conda channel is a directory with certani subdirectories). E.g. 

```
output
 - osx-arm64
   - repodata.json
   - foo-1.2.3-hash.conda
 - linux-64 
   - repodata.json
 - noarch
   - repodata.json
```

So you can also point to `/path/to/output` in your `pixi.toml` and pixi should find the local packages.
